### ISSUE TITLE
Digitizer/stylus not working on Lenovo 500e 2nd gen (81MC)

### ISSUE THREAD

**Issue by dewitte-77** [MrChromebox/firmware#419]
These devices have a garaged stylus that works in chromeos and is  based on the Wacom technology (confirmed by opening the device and checking the digitizer chip -> WACOM W9013 045KM408).

Linux is working  with multiple distro's tested (arch, ubuntu 22.10, opensuse 15.4 with latest firmware 4.18.2) and supporting almost all the hardware except for this stylus.
The touch screen itself is working when using fingers but when using the stylus there is no response at all.

I also cannot find any "unknown" device that might represent the digitizer.
Is it possible that this device is not activated through the firmware?

What information could help to diagnose?

```
 ChromeOS Device Firmware Utility Script [2022-11-29] 
 (c) Mr Chromebox <mrchromebox@gmail.com> 
*********************************************************
**   Device: Lenovo 300e/500e Chromebook 2nd Gen (PHASER360)
** Platform: Intel GeminiLake
**  Fw Type: Full ROM / UEFI
**   Fw Ver: MrChromebox-4.18.2 (11/29/2022)
**    Fw WP: Disabled
*********************************************************
** [WP] 1) Install/Update UEFI (Full ROM) Firmware 
** [WP] 2) Restore Stock Firmware 
**      C) Clear UEFI NVRAM 
*********************************************************
Select a menu option or
R to reboot  P to poweroff  Q to quit 
```

I have reverted back to ChromeOs to get more details of the wacom device.
```
<6>[    1.699604] input: WCOM50C1:00 2D1F:0078 as /devices/pci0000:00/0000:00:16.0/i2c_designware.0/i2c-6/i2c-WCOM50C1:00/0018:2D1F:0078.0001/input/input4
<6>[    1.699833] hid-generic 0018:2D1F:0078.0001: input,hidraw0: I2C HID v1.00 Device [WCOM50C1:00 2D1F:0078] on i2c-WCOM50C1:00


<6>[  647.726962] calling  i2c-WCOM50C1:00+ @ 7584, parent: i2c-6, cb: acpi_subsys_suspend
<6>[  647.727492] call i2c-WCOM50C1:00+ returned 0 after 509 usecs

<6>[  647.761799] calling  i2c-WCOM50C1:00+ @ 7590, parent: i2c-6, cb: acpi_subsys_suspend_late
<6>[  647.761908] call i2c-WCOM50C1:00+ returned 0 after 104 usecs

I: Bus=0018 Vendor=2d1f Product=0078 Version=0100
N: Name="WCOM50C1:00 2D1F:0078"
P: Phys=i2c-WCOM50C1:00
S: Sysfs=/devices/pci0000:00/0000:00:16.0/i2c_designware.0/i2c-6/i2c-WCOM50C1:00/0018:2D1F:0078.0001/input/input4
U: Uniq=
H: Handlers=event4 
B: PROP=2
B: EV=1b
B: KEY=1c03 1 0 0 0 0
B: ABS=d000003
B: MSC=10

2023-01-01T19:38:26.441411Z INFO powerd: [udev.cc(537)] Adding device /sys/devices/pci0000:00/0000:00:16.0/i2c_designware.0/i2c-6/i2c-WCOM50C1:00/0018:2D1F:0078.0001/input/input4 with tags usable_when_laptop usable_when_tablet wakeup wakeup_only_when_usable inhibit role_internal_stylus
2023-01-01T19:38:26.442369Z INFO powerd: [udev.cc(537)] Adding device /sys/devices/pci0000:00/0000:00:16.0/i2c_designware.0/i2c-6/i2c-WCOM50C1:00/0018:2D1F:0078.0001/input/input4/event4 with tags usable_when_laptop usable_when_tablet role_internal_stylus

2023-01-01T19:38:26.565900Z INFO powerd: [input_device_controller.cc(217)] Un-inhibiting /sys/devices/pci0000:00/0000:00:16.0/i2c_designware.0/i2c-6/i2c-WCOM50C1:00/0018:2D1F:0078.0001/input/input4

2023-01-01T19:09:56.842762Z INFO powerd: [input_device_controller.cc(217)] Inhibiting /sys/devices/pci0000:00/0000:00:16.0/i2c_designware.0/i2c-6/i2c-WCOM50C1:00/0018:2D1F:0078.0001/input/input4

dw_i2c: bad counts. hcnt = -23 lcnt = 32
\_SB.PCI0.HDAS.MAXM: Maxim Integrated 98357A Amplifier
\_SB.PCI0.I2C0.H009: WCOM Digitizer at I2C: 01:09
\_SB.PCI0.I2C5.DLG7: Dialog Semiconductor DA7219 Audio Codec address 01ah irq 96
\_SB.PCI0.I2C6.D015: ELAN Touchpad at I2C: 03:15
\_SB.PCI0.I2C6.H02C: Synaptics Touchpad at I2C: 03:2c
\_SB.PCI0.I2C7.D010: ELAN Touchscreen at I2C: 04:10
\_SB.PCI0.I2C7.H020: Synaptics Touchscreen at I2C: 04:20
\_SB.PCI0.I2C7.D039: Raydium Touchscreen at I2C: 04:39
\_SB.PCI0.I2C7.H040: G2TOUCH Touchscreen at I2C: 04:40
\_SB.PCI0.SPI1.S000: SPI Device at SPI: 00
\_SB.PCI0.XHCI.RHUB.HS03: Bluetooth at USB2 port 2
\_SB.PCI0.XHCI.RHUB.HS09: Bluetooth at USB2 port 8

 {
   "componentCategory": "stylus",
   "componentUuid": "phaser360_stylus_2134_3683#3",
   "qualificationStatus": "QUALIFIED",
   "componentFields": {
    "stylus": {
     "name": "WCOM50C1:00 2D1F:0078",
     "path": "",
     "event": "",
     "bus": 0,
     "vendor": 11551,
     "product": 120,
     "version": 0,
     "fwVersion": "",
     "deviceType": "TYPE_STYLUS"
    }
2023-01-01T19:38:26.724742Z NOTICE chromeos-elan-touch-firmware-update-legacy[983]: ELAN0001:00 Updater Firmware: 5640
2023-01-01T19:38:26.746625Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Touch controller reported Product ID: 190.0
2023-01-01T19:38:26.747559Z NOTICE chromeos-elan-touch-firmware-update-legacy[983]: ELAN0001:00 Firmware up to date.
2023-01-01T19:38:26.765550Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Attempting to load FW: '/lib/firmware/elan_i2c_190.0.bin'
2023-01-01T19:38:26.768675Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 (which points to '/opt/google/touch/firmware/190.0_3.0.bin')
2023-01-01T19:38:26.771289Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Hardware product id : 190.0
2023-01-01T19:38:26.773879Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Updater product id  : 190.0
2023-01-01T19:38:26.780771Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Current Firmware: 3.0
2023-01-01T19:38:26.783735Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Updater Firmware: 3.0
2023-01-01T19:38:26.795789Z NOTICE chromeos-elan-touch-firmware-update-legacy[1160]: ELAN0000:00 Firmware up to date.
2023-01-01T19:38:26.919628Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Chassis identifier detected as: PHASER360
2023-01-01T19:38:26.921684Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Platform version detected as: 5
2023-01-01T19:38:26.923885Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Wacom device hardware ID dectected as: 
2023-01-01T19:38:26.927421Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Device product id detected as: 78
2023-01-01T19:38:26.931151Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Attempting to find FW_V2: /lib/firmware/wacom2_firmware.hex
2023-01-01T19:38:26.933258Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Wacom fw v2 not found. Will use v1.
2023-01-01T19:38:26.936556Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Attempting to Load FW: '/lib/firmware/wacom_firmware.hex'
2023-01-01T19:38:26.943089Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 Active firmware version: 5953
2023-01-01T19:38:26.944966Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 New firmware version: 
2023-01-01T19:38:26.950180Z NOTICE chromeos-wacom-touch-firmware-update-legacy[1136]: i2c-6 error: Unable to find new FW version on disk.
```

**Comment by mpeter50** [comment]
I have experienced this with the same model, but haven't got around yet to figure out whats wrong.
I'll follow this issue and try to help if I can.

**Comment by ellyq** [comment]
As mentioned, octopus (GLK baseboard) is missing pen power sequencing.
I tried to fix it and gave one user test rom few months back, but it didn't work. They were worried about bricking their machine, as they didn't have SuzyQ, so they gave up on that functionality.

Neither MrChromebox, CoolStar or me have any GLK-based machines with pen support to test, so someone with such machine would need to build SuzyQ.

Please attach logs from the script. Make sure to remove any private information (such as WiFi AP MAC in dmesg).

```
#!/bin/bash

printf 'Welcome.\nThis script will ask you for sudo password, which is necessary to extract required information.\n'

# Get current username before escalating permissions
user=$USER

# Create directory to store logs
mkdir -p ~/Desktop/debug-logs
cd ~/Desktop/debug-logs

# Download cbmem and mark it as executable
wget https://elly.rocks/tmp/coreboot-development/cbmem
chmod +x cbmem

# Priviledge escalation [!!!]
sudo su <<EOF

# Grab logs and redirect output to files instead of stdout
dmesg >> dmesg.log

if [ ! -f /sbin/lspci ]
then
	printf 'lspci not found. Please install pciutils.'
else
	lspci -vvvnn >> lspci.log
fi

if [ ! -f /bin/lsusb ]
then 
	printf 'lsusb not found. Please install usbutils.'
else
	lsusb -v >> lsusb.log
fi

if [ ! -f /sbin/dmidecode ]
then
	printf 'Dmidecode not found. Please install it'.
else
	dmidecode > dmidecode.log
fi

## Copy ACPI tables
cp /sys/firmware/acpi/tables/DSDT .
cp /sys/firmware/acpi/tables/SSDT* .

# Grab coreboot logs
./cbmem -c > cbmem.log

# Set file permissions for regular user
chown $user:$user *
chmod 755 *

EOF

# Pack logs into archive and remove temporary folder that stores them
cd ..
tar -czf debug-logs.tar.gz debug-logs
rm -r debug-logs

printf 'Log collection done.\nPlease upload "debug-logs.tar.gz" for analysis.\n'
```

**Comment by mpeter50** [comment]
> Neither MrChromebox, CoolStar or me have any GLK-based machines with pen support to test, so someone with such machine would need to build SuzyQ.

I don't have such a cable, but wanted to have one anyways, so I could build one and help with this.
However it might take some time to do so, as last time I tried I had problems with sourcing parts, so if dewitte-77 has one already that would probably be faster. I'll now prioritize this, though.

**Comment by dewitte-77** [comment]
I have ordered the necessary parts to build the cable.
Just for my understanding: do we only need the cable when things go wrong? (for example, the test firmware does not work and we have to use the cable to unbrick)
So I can already execute the script and upload the requested file for you guys to create a test firmware?

**Comment by ellyq** [comment]
Yes please, it will provide crucial information for troubleshooting.

**Comment by MrChromebox** [comment]
one does not need a Suzy-Q cable to debug this. If the device isn't showing up in cbmem under the SSDT, then the ACPI code isn't present for the OS driver to attach

**Comment by dewitte-77** [comment]
[debug-logs.tar.gz](https://github.com/MrChromebox/firmware/files/10376646/debug-logs.tar.gz)
Had to build cbmem because the download link gave errors 
Hope this contains the information you need.

**Comment by MrChromebox** [comment]
looking at the code for the 4.18.x release, it seems that the reset line for the digitizer/pen was not being released prior to detection, so the ACPI wasn't getting added. This was corrected when upstreaming, and should be fixed in the upcoming 4.19 release. If you ping me on the chrultrabook Discord, I can whip up a test build for you to try

**Comment by dewitte-77** [comment]
After installing the test build         
        Version: MrChromebox-4.18.2
        Release Date: 01/10/2023
the Digitizer/stylus is detected by the kernel (tested with arch 6.1.4-arch1-1) and working!
Thank you all for contributing to this solution.
